File: 210-course-schedule-ii/course-schedule-ii.java
class Solution {
    public int[] findOrder(int numCourses, int[][] prerequisites) 
    {
        List<Integer>[] adj = new ArrayList[numCourses];
        // Initialize adjacency list
        for(int i = 0; i < numCourses; i++) 
        adj[i] = new ArrayList<>();

        // Build adjacency list
        for(int i = 0; i < prerequisites.length; i++) 
        {
            int course = prerequisites[i][0];
            int prerequisite = prerequisites[i][1];
            adj[prerequisite].add(course);
        }

        // Build in-degree array
        int[] inDegree = new int[numCourses];
        for(int i = 0; i < numCourses; i++)
        {
            for(int neighbor : adj[i]) 
            inDegree[neighbor]++;
        }

        // Add courses with in-degree 0 to the queue
        Queue<Integer> queue = new LinkedList<>();
        for(int i = 0; i < numCourses; i++) 
        {
            if(inDegree[i] == 0) 
            queue.add(i);
        }

        // Process courses
        int finishedCourses = 0;
        int[] courseList = new int[numCourses];
        while(!queue.isEmpty()) 
        {
            int node = queue.poll();
            courseList[finishedCourses] = node;
            finishedCourses++;

            for(int neighbor : adj[node]) 
            {
                inDegree[neighbor]--;
                if(inDegree[neighbor] == 0) 
                queue.add(neighbor);
            }
        }
        return numCourses == finishedCourses ? courseList : new int[]{};
    }
}

File: 210-course-schedule-ii/course-schedule-ii.cpp
class Solution {
public:
    vector<int> findOrder(int V, vector<vector<int>>& prerequisites) 
    {
        // creating adj list.
        vector<vector<int>> adj(V);

	    // creating indegree vector.
	    vector<int> ind(V,0);

        for(auto x : prerequisites)
        {
            adj[x[1]].push_back(x[0]);
            ind[x[0]]++;
        }
        
        queue<int> q;
        // pushing nodes having zero indegree.
	    for(int i = 0; i < V; i++)        
	    {
	        if(ind[i] == 0)
	        q.push(i);
	    }

        vector<int> order;
	    while(!q.empty())
	    {
	        int node=q.front();
	        q.pop();
            // push in the result
            order.push_back(node);

	        for(int x : adj[node])
	        {
	            ind[x]--;           //decreasing indegree by 1.
	            if(ind[x] == 0)       //if indegree becomes zero. push the node in the queue.
	            q.push(x);
	        }
	    }

	    return order.size() == V ? order : vector<int>();
    }
};